### PR TITLE
Values set by mask and background shorthands should not serialize as "initial"

### PR DESCRIPTION
**Pull request by darinadler**
#### 5d2cf8738277ddd1678dba6ea00157c60f247814
<pre>
Values set by mask and background shorthands should not serialize as &quot;initial&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=251838">https://bugs.webkit.org/show_bug.cgi?id=251838</a>
rdar://problem/105114588

Reviewed by Tim Nguyen.

Needed to expand the function that serializes longhands to cover this case.
Also fixed serialization of the background shorthand to more closely match
the specification and a mistake in the parser that allowed invalid background
shorthand values involving sizes without positions.

* LayoutTests/editing/deleting/paste-with-transparent-background-color-expected.txt:
* LayoutTests/editing/deleting/paste-with-transparent-background-color-live-range-expected.txt:
Updated for progression where we no longer add &quot;background-image: none&quot; as a side effect.

* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-after-set-backgroundSize.html:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style-expected.txt:
* LayoutTests/fast/backgrounds/background-shorthand-with-backgroundSize-style.html:
* LayoutTests/fast/css/remove-shorthand-expected.txt:
* LayoutTests/fast/dom/background-shorthand-csstext-expected.txt:
* LayoutTests/fast/dom/background-shorthand-csstext.html:
Updated to assume background-color will be serialized first, as specified.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-invalid-expected.txt:
Updated to expect PASS for the test of parsing a size without a position.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-valid.html:
Added tests of the longhand values. Removed result that allowed WebKit&apos;s old
incorrect serialization.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-masking/parsing/mask-valid.sub-expected.txt:
Added tests of the longhand values.

* Source/WebCore/css/CSSProperties.json: Moved background-color to the start
of the background shorthand so serialization matches the specification.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::serializeLonghandValue): Added special case for the lists that are
used for the mask and background longhands, because the parser puts
placeholders in there when parsing the shorthands.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Only parse size
immediately after a position. The old code required that it be after a
position, but not immediately after. Removed comment saying this makes
&quot;invalid longhands&quot;. They are now valid, and the serializeLonghandValue
function above makes them serialize correctly.

Canonical link: <a href="https://commits.webkit.org/260157@main">https://commits.webkit.org/260157@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9e49d81841b5848e08ad7377b2316ed0274f0675

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16380 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40158 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116480 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111223 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7619 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99480 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113087 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13439 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41054 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95343 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/28191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82826 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9411 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29535 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10056 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6490 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15564 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49115 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7026 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11610 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Review by nt1m — APPROVED** [review]
Looks good! Can you export the WPT changes and check the cross browser results in the checks tab on the WPT PR?

**Comment by darinadler** [comment]
> Looks good! Can you export the WPT changes and check the cross browser results in the checks tab on the WPT PR?

I exported the changes in https://github.com/web-platform-tests/wpt/pull/38456 but I don’t really know how to check the results for other browsers; need someone to show me how, I guess.

**Comment by Loirooriol** [comment]
It's in the Checks tab: [Chrome](https://github.com/web-platform-tests/wpt/pull/38456/checks?check_run_id=11273810060), [Firefox](https://github.com/web-platform-tests/wpt/pull/38456/checks?check_run_id=11273814475), [Safari](https://github.com/web-platform-tests/wpt/pull/38456/checks?check_run_id=11273779453).

**Comment by nt1m** [comment]
Interesting, Firefox serializes `background-color: transparent` to `background-color: rgba(0, 0, 0, 0)` and doesn't obey the shortest serialization rule for `mask-position` / `background-position` (`0% 0%` vs. `0%`). Worth filing an CSSWG issue on the color part.

Chrome seems to keep serializing with initial which is wrong.

Here's the Firefox diff: https://wpt.fyi/results/css/css-backgrounds/parsing/background-valid.html?diff&filter=ADC&run_id=5146658350825472&run_id=6193847596482560

Chrome diff: https://wpt.fyi/results/css/css-backgrounds/parsing/background-valid.html?diff&filter=ADC&run_id=6280005059870720&run_id=6266588706111488

Seems good to go!

File: Source/WebCore/css/CSSProperties.json [hunk 1]
```diff
@@ -1489,15 +1489,15 @@
         "background": {
             "codegen-properties": {
                 "longhands": [
+                    "background-color",
```
Comment (Loirooriol) [inline Source/WebCore/css/CSSProperties.json:1492]: Why move this? The [spec](https://drafts.csswg.org/css-backgrounds-3/#background) says "canonical order: per grammar", and `backgrfound-color` only appears at the end in the `<final-bg-layer>`.

Comment (nt1m) [reply Source/WebCore/css/CSSProperties.json:1492]: If there's only one background-image, then background-color does appear first. The spec is just saying you can't mention multiple background-colors in the shorthand (which makes sense). 

I think this deserves some clarification?

Comment (Loirooriol) [reply Source/WebCore/css/CSSProperties.json:1492]: Well, even if there is only a single layer, `<bg-layer>` will be omitted but it's still in the grammar, so `background-color` appears last. I agree that "canonical order: per grammar" may not be much clear in these complex cases, but I don't see what backs this:

> Moved background-color to the start of the background shorthand so serialization matches the specification.

Comment (darinadler) [reply Source/WebCore/css/CSSProperties.json:1492]: https://www.w3.org/TR/css-backgrounds-3/#typedef-final-bg-layer lists the color first. That is where I got the idea from. Maybe I am looking in the wrong place.

Comment (darinadler) [reply Source/WebCore/css/CSSProperties.json:1492]: I see your point now.

I’d be happy to move `background-color` back to the end. I suggest we try to resolve this in https://github.com/w3c/csswg-drafts/issues/418 and then change to match at that time. Whoever wrote the WPT test apparently read the specification the same way I did.

Comment (darinadler) [reply Source/WebCore/css/CSSProperties.json:1492]: Note that the WebKit parser does **not** parse a trailing color as a separate final background layer, nor should it. So while the grammar states something like that, it’s not at all consistent with how the values are interpreted. Anyway, I’m willing to change back.

**Pull request merged**